libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -0,0 +1,2751 @@
+      ForAll set b
+        | Lit a <- simplifyTerm $ substituteTerm sub set -> do
+            tell $ Any True
+            pure $ ForAll (Lit a) b
+        | otherwise -> ForAll set <$> goBinder fvs sub b
+      Case t bs
+        | Lit a <- simplifyTerm $ substituteTerm sub t -> do
+            tell $ Any True
+            pure $ runCaseOn a bs $ \x v p -> substPred (singletonEnv x v) p
+        | otherwise -> Case t <$> mapMList (goBinder fvs sub) bs
+      Let t (x :-> p)
+        | all (\n -> count n fvs <= 1) (freeVarSet t) -> do
+            tell $ Any True
+            pure $ substitutePred x t p
+        | count (Name x) (freeVars p) == 0 -> do
+            tell $ Any True
+            pure p
+        | Lit a <- simplifyTerm $ substituteTerm sub t -> do
+            tell $ Any True
+            pure $ unBind a (x :-> p)
+        | otherwise -> Let t . (x :->) <$> go (underBinder fvs x p) (x := t : sub) p
+      Exists k b -> Exists k <$> goBinder fvs sub b
+      Block ps -> fold <$> mapM (go fvs sub) ps
+      Assert es t
+        | Lit b <- simplifyTerm $ substituteTerm sub t -> do
+            tell $ Any True
+            pure $ assert_ es b
+        | otherwise -> pure p
+      DependsOn t t'
+        | Lit {} <- simplifyTerm $ substituteTerm sub t -> do
+            tell $ Any True
+            pure $ TruePred
+        | Lit {} <- simplifyTerm $ substituteTerm sub t' -> do
+            tell $ Any True
+            pure $ TruePred
+        | otherwise -> pure p
+      TruePred -> pure p
+      FalsePred {} -> pure p
+
+simplifyBinder :: Binder univ a -> Binder univ a
+simplifyBinder (x :-> p) = x :-> simplifyPred p
+
+simplifyPred :: IsUniverse univ => Pred univ -> Pred univ
+simplifyPred = \case
+  Assert es t -> assert_ es (simplifyTerm t)
+  Reflect tm f b -> mkReflect (simplifyTerm tm) f (simplifyBinder b)
+  ForAll set b -> ForAll (simplifyTerm set) (simplifyBinder b)
+  DependsOn x y -> DependsOn x y
+  Case t bs -> Case (simplifyTerm t) (mapList simplifyBinder bs)
+  TruePred -> TruePred
+  FalsePred es -> FalsePred es
+  Block ps -> fold (simplifyPreds ps)
+  Let t b -> Let (simplifyTerm t) (simplifyBinder b)
+  Exists k b -> Exists k (simplifyBinder b)
+
+simplifyPreds :: IsUniverse univ => [Pred univ] -> [Pred univ]
+simplifyPreds = go [] . map simplifyPred
+  where
+    go acc [] = reverse acc
+    go _ (FalsePred err : _) = [FalsePred err]
+    go acc (TruePred : ps) = go acc ps
+    go acc (p : ps) = go (p : acc) ps
+
+mkReflect :: (HasSpec univ a, HasSpec univ b) => Term univ a -> (a -> b) -> Binder univ b -> Pred univ
+mkReflect (Lit a) f b = unBind (f a) b
+mkReflect t f b = Reflect t f b
+
+runTerm :: MonadGenError m => Env -> Term univ a -> m a
+runTerm env = \case
+  Lit a -> pure a
+  V v -> findEnv env v
+  App f ts -> do
+    vs <- mapMList (fmap Identity . runTerm env) ts
+    pure $ uncurryList_ runIdentity (sem f) vs
+
+assertPred :: (MonadGenError m, FunctionLike univ) => Env -> Pred univ -> m Property
```

Comment:
This looks really similar to `checkPred`, would it work if `assertPred e p = property <$> checkPred e p`?

Reply:
You wouldn't get the right counterexample unfortunately.

I've been thinking about how to restructure this to make it nicer. Probably one should just go via `GE` and sort it out that way.